### PR TITLE
angular-input-validation

### PR DESCRIPTION
**Pull request by AMSANJEEV28**
- by AMSanjeev

### CODE REVIEW

File: source/_posts/angular-input-validation.md [hunk 1]
```diff
@@ -0,0 +1,442 @@
+---
+title: Input Validation in Angular
+contributor: A M Sanjeev
+avatar: a-m-sanjeev.jpg
+description: Learn how to  create and validate Angular Forms.
```
Comment (bartosz-io) [inline source/_posts/angular-input-validation.md:5]: duplicated spaces - please fix

File: source/_posts/angular-input-validation.md [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+
+_When using ngModel, we must either declare the FormControl as "standalone" in ngModelOptions or specify the name attribute to stop Angular from producing an error._
+
+_Additionally, FormsModule must be added to the array of imports in app.module.ts_
+
+![](https://images.surferseo.art/e406e5f5-d36d-4a3a-b58c-83a0cce44de5.png)
+
+## Validation in Template-driven forms
+
+Angular has a few built-in validators to verify common use cases. If we want the validations to use built-in validators, we would need to apply validation attributes to every form field. These validation attributes are identical to the common HTML5 validation attributes like \`required\`, \`minlength\`, \`pattern\`, etc. Angular offers a few directives to match these characteristics with the validator methods listed in the Angular framework.
```
Comment (bartosz-io) [inline source/_posts/angular-input-validation.md:163]: Please remove the `\` and use ticks to highlight the syntax.

File: source/_posts/angular-input-validation.md [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+    ![](https://images.surferseo.art/e9327bfe-9da1-48d1-b691-05809321d7c3.png)
+3.  Now Open app.module.ts and import FormsModule from '@angular/forms' also call it in imports.
+    
+        import { FormsModule } from '@angular/forms';
+    
+          imports: [
+            BrowserModule,
+            AppRoutingModule,
+            FormsModule
+          ]
```
Comment (bartosz-io) [inline source/_posts/angular-input-validation.md:131]: All the code snippets need to be fixed to use triple ticks ```

File: source/_posts/angular-input-validation.md [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+    Here I've named my component as demo-form.
+    
+
+Our newly created component will now show up once we have completed this. More component files are included in this component as \`.ts, .html, and css\` files.
+
+2.  We'll now access the component.html file and call the component tag we just made. This is done in order to integrate the created component into our application. In the case of our project here, it will be:
+    
+        <app-demo-form></app-demo-form>
+    
+    ![](https://images.surferseo.art/e9327bfe-9da1-48d1-b691-05809321d7c3.png)
```
Comment (bartosz-io) [inline source/_posts/angular-input-validation.md:122]: All the images need to be uploaded.

**Review by bartosz-io — CHANGES_REQUESTED** [review]

File: source/_posts/angular-input-validation.md [hunk 5]
```diff
@@ -0,0 +1,508 @@
...
+    ```
+    
+8.  Now check to see whether it's working by opening your browser to [http://localhost:4200/](http://localhost:4200/)
+    
+    {% img "localhost 4200.png" "localhost 4200" "lazy" %}
+    
+
+We're now ready to begin.
+
+## Template driven forms
```
Comment (bartosz-io) [inline source/_posts/angular-input-validation.md:127]: template-driven

File: source/_posts/angular-input-validation.md [hunk 6]
```diff
@@ -0,0 +1,508 @@
...
+
+In Template Driven Forms, we focus on the actions and validations by using directives and attributes in our template and letting it work behind the scenes. Everything happens in Template. The component class requires very little coding. The logic and controls are not defined in the component class here. It is utilized to build straightforward application forms.
```
Comment (bartosz-io) [inline source/_posts/angular-input-validation.md:129]: template-driven

File: source/_posts/angular-input-validation.md [hunk 7]
```diff
@@ -0,0 +1,508 @@
...
+
+1.  All of the components that are being created are located in the src/app folder. We can add another component inside of this one by using the command given below.
+    
+    ```
+    ng g c component_name
+    ```
```
Comment (bartosz-io) [inline source/_posts/angular-input-validation.md:135]: The snippets should not have additional spaces/tabs. Please remove all.

File: source/_posts/angular-input-validation.md [hunk 8]
```diff
@@ -0,0 +1,508 @@
...
+
+Our newly created component will now show up once we have completed this. More component files are included in this component as `.ts`, `.html`, and `css` files.
+
+2.  We'll now access the component.html file and call the component tag we just made. This is done in order to integrate the created component into our application. In the case of our project here, it will be:
+    
+    ```
+    <app-demo-form></app-demo-form>
+    ```
+    
+    {% img "appcomponent.png" "Add codes to appcomponent.html" "lazy" %}
```
Comment (bartosz-io) [inline source/_posts/angular-input-validation.md:148]: The same here: additional spaces/tabs. Please remove all.

File: source/_posts/angular-input-validation.md [hunk 9]
```diff
@@ -0,0 +1,508 @@
...
+
+Every time a FormControl's value changes by performing validation, Angular displays a list of validation problems. The status is **valid** if the list is empty; otherwise, it is **invalid**.
+
+Now, let's say we wanted to add the following validations:
+
+1.  The name field has the required attribute; if this field is left blank, we want to show some error message.
+    
+2.  The minimum and maximum character lengths for the value in the Name field should be 5 and 30, respectively. if not followed, we want to show validation messages.
+    
+
```
Comment (bartosz-io) [inline source/_posts/angular-input-validation.md:204]: Too many new lines...

File: source/_posts/angular-input-validation.md [hunk 10]
```diff
@@ -0,0 +1,508 @@
...
+But how will we know whether or not the contents have been imported?
+
+*   Open the browser and inspect the input tag for this. You'll notice that a class called ngcontent and a few other ng classes have been added, indicating that Angular has identified our demo-form.
+    
+    {% img "ngcontent.png" "ng classes in form" "lazy" %}
+
+Now that our form has been submitted, the submit method and a method called form method should both be called.
+
+*   We will do this by including the following code in our `form` tag.
```
Comment (bartosz-io) [inline source/_posts/angular-input-validation.md:217]: Too many spaces.

File: source/_posts/angular-input-validation.md [hunk 11]
```diff
@@ -0,0 +1,508 @@
...
+    _Note: When we submit and inspect our form, we will see that callback functions have been called._
+    
+    {% img "inspectform.png" "callback functions in form" "lazy" %}
+    
+    Here we can see details about the field's state, including whether it has been **touched** or not, whether it is **clean** or **dirty**, and information about all of FormGroup's properties (**valid**, **invalid**, **pending**, etc.). These are some of the things that we need to keep in mind when implementing form validation.
+    
+
+As we can see from the screenshot above, there were numerous objects, such as **valid** and **invalid**, when we examined our form. We're going to alter that now and use those objects as leverage. We are expected to include a template variable for that.
+
+1.  For this, we are going to add `#variable ="ngModel"` to our input fields. Here I've named our `variable` as `name`.
```
Comment (bartosz-io) [inline source/_posts/angular-input-validation.md:271]: Too many spaces.

File: source/_posts/angular-input-validation.md [hunk 12]
```diff
@@ -0,0 +1,508 @@
...
+    required
+    ngModel
+    />
+    </div>
+
+    <button type="submit" class="btn btn-success">Submit</button>
+    </form>
+    </div>
+
+    ```
```
Comment (bartosz-io) [inline source/_posts/angular-input-validation.md:325]: The formatting of that snippet is totally broken. Please fix it.

File: source/_posts/angular-input-validation.md [hunk 13]
```diff
@@ -0,0 +1,508 @@
...
+interface ValidatorFn {
+(control: AbstractControl): ValidationErrors | null
+}
+```
+
+The ValidationErrors object must include at least one key-value pair:
+
+```
+type ValidationErrors = {
+[key: string]: any;
```
Comment (bartosz-io) [inline source/_posts/angular-input-validation.md:346]: Missing tab.

File: source/_posts/angular-input-validation.md [hunk 14]
```diff
@@ -0,0 +1,508 @@
...
+{% img "validation2.png" "Angular Demo Form with Validations" "lazy" %}
+
+
+## The custom validator in Template-driven forms
+
+There are use cases that the built-in validators can't always help us validate. Then we must develop a unique validator function. The code listed below can be used to define a validator function that implements the ValidatorFn interface.
+
+```
+interface ValidatorFn {
+(control: AbstractControl): ValidationErrors | null
```
Comment (bartosz-io) [inline source/_posts/angular-input-validation.md:338]: Missing tab.

File: source/_posts/angular-input-validation.md [hunk 15]
```diff
@@ -0,0 +1,508 @@
...
+<label for="name">Name</label> 
+<input type="text" class="form-control" id="name" 
+[formControlName]="name"> 
+</div> 
+<div class="form-group"> 
+<label for="username">Username</label> 
+<input type="text" class="form-control" id="username" 
+[formControlName]="username"> 
+</div> 
+<form>
```
Comment (bartosz-io) [inline source/_posts/angular-input-validation.md:411]: Formatting.

File: source/_posts/angular-input-validation.md [hunk 16]
```diff
@@ -0,0 +1,508 @@
...
+```
+registrationForm = new FormGroup({ 
+'name': new FormControl('Enter name', [ 
+Validators.required, 
+Validators.minLength(5), 
+Validators.maxLength(30) 
+]), 
+'username': new FormControl('', Validators.required), 
+}) 
+```
```
Comment (bartosz-io) [inline source/_posts/angular-input-validation.md:443]: Formatting.

File: source/_posts/angular-input-validation.md [hunk 17]
```diff
@@ -0,0 +1,508 @@
...
+</div>
+</div>
+
+<div class="form-group">
+<label for="username">Username</label>
+<input type="text" class="form-control" id="username"
+[formControlName]="username">
+</div>
+
+<form>
```
Comment (bartosz-io) [inline source/_posts/angular-input-validation.md:481]: Formatting.

File: source/_posts/angular-input-validation.md [hunk 18]
```diff
@@ -0,0 +1,508 @@
...
+Validators.required, 
+Validators.minLength(5), 
+Validators.maxLength(30) 
+]), 
+'email': new FormControl('', [ 
+Validators.required, 
+UserRegistrationFormValidators.emailShouldBeValid 
+]), 
+})
+```
```
Comment (bartosz-io) [inline source/_posts/angular-input-validation.md:500]: Formatting.

**Review by bartosz-io — CHANGES_REQUESTED** [review]

**Comment by AMSANJEEV28** [comment]
Done

**Review by bartosz-io — APPROVED** [review]

**Pull request merged**